text/0000-basic-template-helpers.md

```diff
@@ -0,0 +1,86 @@
+- Start Date: 2018-10-13
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# (RFC title goes here)
+
+## Summary
+
+Add new built-in template helpers to perform basic operations in templates, identical to the
+helpers in [ember-truth-helpers](https://github.com/jmurphyau/ember-truth-helpers)
+
+## Motivation
+
+It is a very common necessity to almost every Ember app to perform certain operations like compare
+by equality, negate a value, or perform boolean operations, and often the most convenient place to
+do it is in the templates.
+Because of that, `ember-truth-helpers` is probably the single most installed addon either directly
+or transitively in the entire ecosystem.
+
+The fact that this addon is so popular is very telling and Ember should consider to move into the
+core of the templating engine at least the most popular helpers.
+
+One of the shocking moments developers that are familiar with React, Vue or Angular have when trying Ember
+is that they cannot, at least out of the box, perform the most basic comparisons and boolean operations
+they are so used to in JSX or Vue/Angular templates.
+
+Furthermore, if the [RFC #367](https://github.com/emberjs/rfcs/pull/367) eventually gets merged, these
+helpers are the perfect candidates to create adoption friction in the community because of how
+pervasive its usage it in so many templates.
+
+Albeit less important, implementing these helpers in the core _could_ enable clever trickery in the
+Glimmer VM, like evaluating them at compile time when the arguments are constant.
```

Comment:
Another example would be to short circuit in helpers like `eq`, `and`, and `or`. Today’s version of ember-truth-helpers _always_ evaluates all of the arguments (eg Ember.get’s them), but if implemented in the VM we could avoid evaluation of all arguments after we know the result. 

Take the following example:

```hbs
{{#if (and first second third)}}

{{/if}}
```

Today, we evaluate all of those args, but if implemented in the VM we can avoid evaluating `third` if `second` were falsey.

Reply:
Somewhat related to the `and` semantics: I just saw that the current implementation follows the JS semantics in that it does not necessarily return a strict boolean value, but rather the last argument (if evaluating to a truthy value) or the last falsy argument (if evaluating to false). So `(and "a" "b")` would evaluate to `"b"` (similar to `'a' && 'b'`).

I am not entirely sure I would copy over these JS semantics to hbs-land. At least it was a bit unexpected for me, as I have never used `and` or `or` outside of an `{{#if}}` block. But that can be discussed. If these semantics should remain, this could maybe be pointed out in the RFC, for guiding the future implementation in core...

Reply:
IMHO the current behavior makes sense. It would be very limiting to coerce values to a boolean. I've used the helper that way.